services/offline/src/lib/online-status-message.tsx

```diff
@@ -1,39 +1,64 @@
 import React, { ReactElement, ReactNode, useContext, useState } from 'react'
-import { OnlineStatusMessageContextAPI } from '../types'
 
-const defaultApi: OnlineStatusMessageContextAPI = {
-    onlineStatusMessage: undefined,
-    setOnlineStatusMessage: () => undefined,
-}
-
-const OnlineStatusMessageContext =
-    React.createContext<OnlineStatusMessageContextAPI>(defaultApi)
+type SetOnlineStatusMessage = (message: ReactNode) => void
 
-export const useOnlineStatusMessage = (): OnlineStatusMessageContextAPI => {
-    const { onlineStatusMessage, setOnlineStatusMessage } =
-        useContext<OnlineStatusMessageContextAPI>(OnlineStatusMessageContext)
-
-    return {
-        onlineStatusMessage,
-        setOnlineStatusMessage,
-    }
-}
+// 'get' and 'set' contexts are separated so 'setter' consumers that don't
+// actually need the value don't have to rerender when the value changes:
+const OnlineStatusMessageValueContext = React.createContext<ReactNode>(null)
+const SetOnlineStatusMessageContext =
+    React.createContext<SetOnlineStatusMessage>(() => undefined)
 
 export const OnlineStatusMessageProvider = ({
     children,
 }: {
     children: ReactNode
 }): ReactElement => {
-    const [onlineStatusMessage, setOnlineStatusMessage] = useState<ReactNode>()
+    const [onlineStatusMessage, setOnlineStatusMessage] =
+        useState<ReactNode>(null) // note: not undefined
 
     return (
-        <OnlineStatusMessageContext.Provider
-            value={{
-                onlineStatusMessage,
-                setOnlineStatusMessage,
-            }}
-        >
-            {children}
-        </OnlineStatusMessageContext.Provider>
+        <OnlineStatusMessageValueContext.Provider value={onlineStatusMessage}>
+            <SetOnlineStatusMessageContext.Provider
+                value={setOnlineStatusMessage}
+            >
+                {children}
+            </SetOnlineStatusMessageContext.Provider>
+        </OnlineStatusMessageValueContext.Provider>
     )
 }
+
+export const useOnlineStatusMessageValue = () => {
+    const onlineStatusMessage = useContext(OnlineStatusMessageValueContext)
+
+    // note: value is initialized to `null` in provider, not undefined
+    if (onlineStatusMessage === undefined) {
+        throw new Error(
+            'useOnlineStatusMessageValue must be used within an OnlineStatusMessageProvider'
+        )
+    }
+
+    return onlineStatusMessage
+}
+
+export const useSetOnlineStatusMessage = () => {
+    const setOnlineStatusMessage = useContext(SetOnlineStatusMessageContext)
+
+    if (setOnlineStatusMessage === undefined) {
+        throw new Error(
+            'useSetOnlineStatusMessage must be used within an OnlineStatusMessageProvider'
+        )
+    }
```

Comment:
I think ```setOnlineStatusMessage``` will be the default value of the context - eg. ```() => undefined```` and not undefined if there's no context in the tree?

Reply:
Ah true - I admit that these checks are a bit of an automatic habit 🤦‍♂️ I'll remove them as you suggested in your other comment